### PR TITLE
Introduce codegen for client code - Part 1. #37

### PR DESCRIPTION
**Pull request by kobylynskyi**
Initial work on a code generator for graphql java client.

What's still pending and will be addressed in future PRs:
* Test codegen with customTypesMapping and scalars.
* More customizations:
  * ...
* Gradle/Maven plugins support.
* Write new `example` modules for Gradle/Maven plugins.
* Builder support for `Input` classes that will allow users to easily construct requests. - #10

Any suggestions/ideas/advice will be very appreciated!

### CODE REVIEW

**Comment by kobylynskyi** [comment]
@joffrey-bion, could you please review this PR so that I can merge this one?
Will address all issues (mentioned in the description) in the next Pull Requests.
Thanks

File: src/main/java/com/kobylynskyi/graphql/codegen/model/request/GraphQLRequestSerializer.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.stream.Collectors;
+
+public class GraphQLRequestSerializer {
+
+    public static String serialize(GraphQLRequest graphQLRequest) {
+        if (graphQLRequest == null || graphQLRequest.getRequest() == null) {
+            return null;
+        }
+        StringBuilder builder = new StringBuilder();
+        builder.append("{\"query\":\"");
```
Comment (joffrey-bion) [inline src/main/java/com/kobylynskyi/graphql/codegen/model/request/GraphQLRequestSerializer.java:15]: I believe the JSON construction here should be separate from building the query string (and maybe use a library although this is debatable).
GraphQL requests actually expect the `operationName`, `query`, and `variables` JSON fields AFAIR, maybe we should include that as well.

File: src/main/java/com/kobylynskyi/graphql/codegen/mapper/FieldDefinitionToParameterMapper.java [hunk 2]
```diff
@@ -24,18 +28,74 @@
...
+    /**
+     * Map GraphQL's FieldDefinition to a Freemarker-understandable format of parameter
+     *
+     * @param mappingConfig  Global mapping configuration
+     * @param fieldDef       GraphQL field definition
+     * @param parentTypeName Name of the parent type
+     * @param typeNames      Names of all GraphQL types
+     * @return Freemarker-understandable format of parameter (field)
+     */
+    private static ProjectionParameterDefinition mapProjectionFields(MappingConfig mappingConfig, FieldDefinition fieldDef, String parentTypeName, Set<String> typeNames) {
```
[the remark this answers is not in the document]
Comment (kobylynskyi) [reply src/main/java/com/kobylynskyi/graphql/codegen/mapper/FieldDefinitionToParameterMapper.java]: Fixed. Thanks!

File: src/main/java/com/kobylynskyi/graphql/codegen/mapper/FieldDefinitionToParameterMapper.java [hunk 3]
```diff
@@ -24,18 +28,74 @@
...
 
+    /**
+     * Map GraphQL's FieldDefinition to a Freemarker-understandable format of parameter
+     *
+     * @param mappingConfig  Global mapping configuration
+     * @param fieldDef       GraphQL field definition
+     * @param parentTypeName Name of the parent type
+     * @return Freemarker-understandable format of parameter (field)
+     */
+    private static ParameterDefinition mapFields(MappingConfig mappingConfig, FieldDefinition fieldDef, String parentTypeName) {
```
[the remark this answers is not in the document]
Comment (kobylynskyi) [reply src/main/java/com/kobylynskyi/graphql/codegen/mapper/FieldDefinitionToParameterMapper.java]: Fixed. Thanks!

File: src/main/java/com/kobylynskyi/graphql/codegen/model/request/GraphQLRequestSerializer.java (hunk 1 above)
Comment (kobylynskyi) [reply src/main/java/com/kobylynskyi/graphql/codegen/model/request/GraphQLRequestSerializer.java:15]: I know for sure that `operationName`, `query`, `variables` is not mandatory.
The following query works on one of my projects:
```json
{"query":"query { eventsByIds(ids: [ "4", "5", "6" ]) { id } }"}
```
But maybe some other graphql server will expect these parameters. So I will include them. I will take a note and make a change in the upcoming PR. Thanks!

**Pull request merged**